ui/desktop/src/components/context_management/ContextManager.tsx

```diff
@@ -0,0 +1,181 @@
+}
+
+interface ContextManagerActions {
+  handleAutoCompaction: (
+    messages: Message[],
+    setMessages: (messages: Message[]) => void,
+    append: (message: Message) => void,
+    setAncestorMessages?: (messages: Message[]) => void
+  ) => Promise<void>;
+  handleManualCompaction: (
+    messages: Message[],
+    setMessages: (messages: Message[]) => void,
+    append?: (message: Message) => void,
+    clearAlerts?: () => void,
+    setAncestorMessages?: (messages: Message[]) => void
+  ) => Promise<void>;
+  hasCompactionMarker: (message: Message) => boolean;
+}
+
+// Create the context
+const ContextManagerContext = createContext<
+  (ContextManagerState & ContextManagerActions) | undefined
+>(undefined);
+
+// Create the provider component
+export const ContextManagerProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
+  const [isCompacting, setIsCompacting] = useState<boolean>(false);
+  const [compactionError, setCompactionError] = useState<string | null>(null);
+
+  const performCompaction = useCallback(
+    async (
+      messages: Message[],
+      setMessages: (messages: Message[]) => void,
+      append: (message: Message) => void,
+      _isManual: boolean = false,
+      setAncestorMessages?: (messages: Message[]) => void
+    ) => {
+      setIsCompacting(true);
+      setCompactionError(null);
+
+      try {
+        // Get the summary from the backend
+        const summaryResponse = await manageContextFromBackend({
+          messages: messages,
+          manageAction: 'summarize',
+        });
+
+        // Convert API messages to frontend messages
+        const convertedMessages = summaryResponse.messages.map((apiMessage) => {
+          const isCompactionMarker = apiMessage.content.some(
+            (content) => content.type === 'summarizationRequested'
+          );
+
+          if (isCompactionMarker) {
+            // show to user but not model
+            return convertApiMessageToFrontendMessage(apiMessage, true, false);
+          }
+
+          // show to model but not user
+          return convertApiMessageToFrontendMessage(apiMessage, false, true);
+        });
+
+        // Store the original messages as ancestor messages so they can still be scrolled to
+        if (setAncestorMessages) {
+          const ancestorMessages = messages.map((msg) => ({
+            ...msg,
+            display: true,
+            sendToLLM: false,
+          }));
+          setAncestorMessages(ancestorMessages);
+        }
+
+        // Replace messages with the server-provided messages
+        setMessages(convertedMessages);
+
+        // Automatically submit the continuation message to continue the conversation
+        // This should be the third message (index 2) which contains the "I ran into a context length exceeded error..." text
+        const continuationMessage = convertedMessages[2];
+        if (continuationMessage) {
+          setTimeout(() => {
+            append(continuationMessage);
```

Comment:
TODO here - look into putting the original user message in if context ran out entirely